src/callbacks/callbacks.jl

```diff
@@ -347,3 +347,271 @@ function reset_graceful_exit(output_dir)
+            # Return "nothing" if "reduction_time" is not in the YAML block
+            #
+            # We also normalize everything to lowercase, so that can accept "max" but
+            # also "Max"
+            reduction_time_yaml =
+                lowercase(get(yaml_diag, "reduction_time", "nothing"))
+
+            if !haskey(ALLOWED_REDUCTIONS, reduction_time_yaml)
+                error("reduction $reduction_time_yaml not implemented")
+            else
+                reduction_time_func, pre_output_hook! =
+                    ALLOWED_REDUCTIONS[reduction_time_yaml]
+            end
+
+            writer_ext = lowercase(get(yaml_diag, "writer", "nothing"))
+
+            if !haskey(ALLOWED_WRITERS, writer_ext)
+                error("writer $writer_ext not implemented")
+            else
+                writer = ALLOWED_WRITERS[writer_ext]
+            end
+
+            haskey(yaml_diag, "period") ||
+                error("period keyword required for diagnostics")
+
+            period_seconds = time_to_seconds(yaml_diag["period"])
+
+            if isnothing(output_name)
+                output_short_name = CAD.descriptive_short_name(
+                    CAD.get_diagnostic_variable(short_name),
+                    period_seconds,
+                    reduction_time_func,
+                    pre_output_hook!,
+                )
+            end
+
+            if isnothing(reduction_time_func)
+                compute_every = period_seconds
+            else
+                compute_every = :timestep
+            end
+
+            return CAD.ScheduledDiagnosticTime(
+                variable = CAD.get_diagnostic_variable(short_name),
+                output_every = period_seconds,
+                compute_every = compute_every,
+                reduction_time_func = reduction_time_func,
+                pre_output_hook! = pre_output_hook!,
+                output_writer = writer,
+                output_short_name = output_short_name,
+            )
+        end
+        return ret_value
+    end
+
+    # Flatten the array of arrays of diagnostics
+    diagnostics = vcat(diagnostics_ragged...)
+
+    if parsed_args["output_default_diagnostics"]
+        t_end = time_to_seconds(parsed_args["t_end"])
+        return [
+            CAD.default_diagnostics(
+                atmos_model,
+                t_end;
+                output_writer = netcdf_writer,
+            )...,
+            diagnostics...,
+        ],
+        writers
+    else
+        return collect(diagnostics), writers
+    end
+end
+
+function get_diagnostics_cb(
+    Y,
+    p,
+    t_start,
+    config,
+    sim_info,
+    atmos,
+    params,
+    spaces,
+)
+    (; comms_ctx, parsed_args) = config
+
+    s = @timed_str begin
+        diagnostics, writers = get_diagnostics(parsed_args, atmos, spaces)
+    end
+    @info "initializing diagnostics: $s"
+
+    if length(diagnostics) > 0
+        @info "Computing diagnostics:"
+    else
+        return (), ()
```

Comment:
Can please you document the expected return objects/types for this function?

Reply:
I actually had to fix this because it failed in the empty case, I switched to NamedTuples, so it should also be clearer.